sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -210,7 +214,28 @@ private[sbtplugin] object ScalaJSPluginInternal {
             }
 
             Set(output, sourceMapFile)
-          } (realFiles.toSet)
+          }
+
+          val useChanged = (scalaJSUseMainModuleInitializerTask in key)
+            .previous
+            .exists(_ != (scalaJSUseMainModuleInitializerTask in key).value)
```

Comment:
In theory, this check should not be necessary, as a change in `scalaJSUseMainModuleInitializer` implies a change in `scalaJSModuleInitializers`. 

However, in practice, if I remove this call, the `previous` value of `scalaJSModuleInitializers` is not written correctly to the `previous` file. It's as if looking at the `.previous` of `scalaJSUseMainModuleInitializerTask` has a side-effect that makes `scalaJSModuleInitializers` be written correctly.

This is very weird, and I don't know how else to solve the problem.

Reply:
Have you considered dumping all this in a file instead? Then it will also work across sbt invocations.

Reply:
Do you mean manually writing to a file?

Sbt's `.previous` already works across invocations, it reads the serialized JSON of the previous value from a file called  `target/streams/<key>/<value>/previous`.

Reply:
> Sbt's .previous already works across invocations, it reads the serialized JSON of the previous value from a file called target/streams/<key>/<value>/previous.

Ah, OK. In that case I guess we can use this. Although the additional keys are a bit hairy IMHO.

I would like you to investigate why `scalaJSUseMainModuleInitializer` needs to be handled explicitly. It should not be special. If that breaks, an custom setup to generate `scalaJSModuleInitializers` might also break.

Reply:
> 
> 
> In theory, this check should not be necessary, as a change in `scalaJSUseMainModuleInitializer` implies a change in `scalaJSModuleInitializers`.
> 
> However, in practice, if I remove this call, the `previous` value of `scalaJSModuleInitializers` is not written correctly to the `previous` file. It's as if looking at the `.previous` of `scalaJSUseMainModuleInitializerTask` has a side-effect that makes `scalaJSModuleInitializers` be written correctly.
> 
> This is very weird, and I don't know how else to solve the problem.

Is this still true, even though everything was now moved to separate tasks with the fingerprints?

Reply:
What if you always store the `.value` of the fingerprints in `val`s, and the `.previous` in `val`s as well? Hiding them in local `def`s and in the lambdas passed to `.exists` might confuse sbt's macros that analyze the dependencies. In fact I am surprised it doesn't flag those usages as downright illegal.

Reply:
> I would like you to investigate why `scalaJSUseMainModuleInitializer` needs to be handled explicitly. It should not be special.

I've looked into it,  and removing the explicit handling of the setting works now. I don't know why it didn't work before. I'm unable to replicate any of the weird behavior I had before, so I think I must've done something wrong in my testing setup back then.

In any case, this is fixed as of 49f4a0280d58ccbed98a5ad688eb223f4b2536ed. Changes in `scalaJSUseMainModuleInitializer` are indeed captured in `scalaJSModuleInitializers`, as expected.

Reply:
> Hiding them in local defs and in the lambdas passed to .exists might confuse sbt's macros that analyze the dependencies.

My initial implementation actually stored the comparison (with `.exists`) in a `val`. I eventually changed it to `def` so that the short-circuiting `||` operator could save us a file read in some scenarios.

Based on all my testing, it seems like it works, and sbt doesn't complain :man_shrugging: